### PR TITLE
Improve type inference

### PR DESCRIPTION
**Pull request by definelicht**
Use a more intelligent type inference system (would still be better to rely on something external for this?).

### CODE REVIEW

**Comment by tbennun** [comment]
have you looked at `typed-ast`? might give us what we need?

File: dace/codegen/cppunparse.py [hunk 1]
```diff
@@ -811,43 +811,16 @@ def _Num(self, t, infer_type=False):
...
+            self.write(repr_n.replace("inf", INFSTR), infer_type)
+            # If the number has a type, use it
+            if infer_type:
+                return dace.dtypes.typeclass(type(t.n))
             else:
-                # Substitute overflowing decimal literal for AST infinities.
-                self.write(repr_n.replace("inf", INFSTR))
-
-            if repr_n.startswith("-"):
-                self.write(")")
```
Comment (tbennun) [inline dace/codegen/cppunparse.py:850]: what happened to this?

File: dace/dtypes.py [hunk 2]
```diff
@@ -260,159 +260,43 @@ def __repr__(self):
...
+    # Both are integers
+    if numpy.issubdtype(lhs_, numpy.integer) and numpy.issubdtype(
+            rhs_, numpy.integer):
+        # If one byte width is larger, use it
+        if lhs_.itemsize > rhs_.itemsize:
+            return lhs
+        elif lhs_.itemsize < rhs_.itemsize:
+            return rhs
+        # Sizes are the same
+        if numpy.issubdtype(lhs_, numpy.unsignedinteger):
```
Comment (tbennun) [inline dace/dtypes.py:284]: what about the other side of this?

File: dace/codegen/cppunparse.py (hunk 1 above)
Comment (definelicht) [reply dace/codegen/cppunparse.py:850]: It's above, just diff makes it look like it was deleted and re-added

File: dace/dtypes.py (hunk 2 above)
Comment (definelicht) [reply dace/dtypes.py:284]: What do you mean?

Comment (tbennun) [reply dace/dtypes.py:287]: if `rhs_` is unsigned but `lhs_` isn't.

Comment (definelicht) [reply dace/dtypes.py:287]: Then we have to return `rhs_`, which is what we do

Comment (definelicht) [reply dace/dtypes.py:287]: It's in the comments :-)

**Pull request merged**